rna/unafold/doc/hybrid-2s.pl.1.in
.TH HYBRID-2S.PL 1 "@DATE@" "@PACKAGE_NAME@ @PACKAGE_VERSION@" "User Commands"
.SH NAME
hybrid-2s.pl \- hybridize two RNA or DNA sequences
.SH SYNOPSIS
.B hybrid-2s.pl
.RI [ OPTION ]...
.I FILE1 FILE2
.SH DESCRIPTION
\fBhybrid-2s.pl\fR computes a minimum energy hybridization of the sequences in \fIFILE1\fR and \fIFILE2\fR and evaulates its enthalpy and entropy.  It then calculates the free energy at each temperature in a range and writes a \fI.dG\fR file.  \fBhybrid-2s.pl\fR also makes a copy of the \fI.ext\fR file for each temperature, so that it can act as a replacement for \fBhybrid\fR, for example in \fBhybrid2.pl\fR.
.SH OPTIONS
.TP
.BR -n ", " --NA =RNA|DNA
sets nucleic acid type to RNA or DNA.  Default is RNA.
.TP
.BR -t ", " --tmin =\fItemp
sets minimum temperature to \fItemp\fR.  Default is 37.
.TP
.BR -i ", " --tinc =\fItemp
sets temperature increment to \fItemp\fR.  Default is 1.
.TP
.BR -T ", " --tmax =\fItemp
sets maximum temperature to \fItemp\fR.  Default is 37.
.TP
.BR -N ", " --sodium =\fIvalue
sets Sodium ion concentration to \fIvalue\fR molar.  Default is 1.
.TP
.BR -M ", " --magnesium =\fIvalue
sets Magnesium ion concentration to \fIvalue\fR molar.  Default is 0.
.TP
.BR -p ", " --polymer
use salt corrections for polymers instead of oligomers (the default).
.TP
.BR -r ", " --prohibit =\fIi\fR,\fIj\fR[,\fIk\fR]
prohibit all basepairs in the helix from \fIi\fR,\fIj\fR to \fIi+k-1\fR,\fIj-k+1\fR.  If \fIj\fR is 0, prohibit bases \fIi\fR to \fIi+k-1\fR of the first sequence from pairing at all; if \fIi\fR is 0, prohibit bases \fIj\fR to \fIj-k+1\fR of the second sequence from pairing at all.  \fIk\fR defaults to 1 if not specified.
.TP
.BR -f ", " --force =\fIi\fR,\fIj\fR[,\fIk\fR]
force all basepairs in the helix from \fIi\fR,\fIj\fR to \fIi+k-1\fR,\fIj-k+1\fR.  If \fIj\fR is 0, forces bases \fIi\fR to \fIi+k-1\fR of the first sequence to be double-stranded; if \fIi\fR is 0, forces bases \fIj\fR to \fIj-k+1\fR of the second sequence to be double-stranded.  \fIk\fR defaults to 1 if not specified.
.TP
.BR -E ", " --energyOnly
skips computation of probabilities and outputs only \fIprefix.dG\fR and \fIprefix.run\fR.
.TP
.BR -I ", " --noisolate
prohibits all isolated basepairs.  Isolated basepairs are helices of length 1; that is, they do not stack on another basepair on either side.  (See also the \fB--prefilter\fR and \fB--nopostfilter\fR options below.)
.TP
.BR -c ", " --constraints =\fIfile
reads a list of constraints from \fIfile\fR.  Constraints must be in the form "P \fIi\fR \fIj\fR \fIk\fR" or "F \fIi\fR \fIj\fR \fIk\fR".  These are equivalent to specifying "--prohibit=\fIi\fR,\fIj\fR,\fIk\fR" and "--force=\fIi\fR,\fIj\fR,\fIk\fR", respectively.  If \fIfile\fR is not specified, it defaults to \fIprefix.aux\fR.
.TP
.BR -b ", " --basepairs =\fIfile
reads a list of allowable helices from \fIfile\fR.  Each helix consists of three whitespace-delimited numbers which specify the starting basepair and the length of the helix.  When this option is used, all basepairs except those in \fIfile\fR are prohibited from forming.
.TP
.BR --temperature =\fItemp
sets the temperature at which the minimum-energy hybridization is computed.  Default is 37.
.SH OBSCURE OPTIONS
.TP
.B --allpairs
allows basepairs to form between any two nucleotides.  When \fB--allpairs\fR is not specified, only Watson-Crick and wobble basepairs are allowed.
.TP
.BR --maxloop =\fIsize
sets the maximum size of bulge/interior loops to \fIsize\fR.  Default is 30.
.TP
.B --nodangle
removes single-base stacking from consideration.
.TP
.BR --prefilter =\fIvalue1\fR[,\fIvalue2\fR]
Sets the prefilter to filter out all basepairs except those in groups of \fIvalue2\fR adjacent basepairs of which \fIvalue1\fR can form.  \fIvalue2\fR is the same as \fIvalue1\fR if unspecified.  Default is 2 of 2.  (See also the \fB--noisolate\fR option above.)
.SH ENVIRONMENT
.IP UNAFOLDDAT
an alternate location from which to read the energy rules.  The default energy rules can be overridden with files in the current directory or in the directory pointed to by UNAFOLDDAT.  \fBhybrid-2s.pl\fR looks for each file first in the current directory, then in the directory specified by UNAFOLDDAT and last in \fI/usr/local/share/unafold\fR (or wherever the energy rules were installed).
.SH REFERENCES
@REFERENCES@
.SH "SEE ALSO"
.BR hybrid (1),
.BR hybrid-min (1),
.BR hybrid-ss-2s.pl (1),
.BR hybrid2-2s.pl (1)
.SH AUTHORS
Nick Markham <markhn@rpi.edu> and Michael Zuker <zukerm@rpi.edu>
.SH COPYRIGHT
Copyright (c) 2006-2009, Rensselaer Polytechnic Institute.
.SH AVAILABILITY
UNAFold is available from http://dinamelt.bioinfo.rpi.edu/unafold/.  Commercial use requires a license; see http://dinamelt.bioinfo.rpi.edu/commercial.php.


rna/unafold/scripts/hybrid-2s.pl
#! /opt/local/bin/perl -w

use strict;
use warnings;
use Getopt::Long;
Getopt::Long::Configure 'gnu_getopt', 'no_auto_abbrev', 'no_ignore_case';

use constant R => .0019872;

sub version ($) {
    print "$_[0] (UNAFold) 3.8\n";
    print "By Nicholas R. Markham and Michael Zuker\n";
    print "Copyright (C) 2006\n";
    print "Rensselaer Polytechnic Institute\n";
    print "Troy, NY 12810-3590 USA\n";
    exit;
}

sub usage () {
    print <<EOF;
Usage: hybrid-2s.pl [options] file1 file2

Options:
-n, --NA=(RNA | DNA) (defaults to RNA)
-t, --tmin=<minimum temperature> (defaults to 0)
-i, --tinc=<temperature increment> (defaults to 1)
-T, --tmax=<maximum temperature> (defaults to 100)
-N, --sodium=<[Na+] in M> (defaults to 1)
-M, --magnesium=<[Mg++] in M> (defaults to 0)
-p, --polymer
-h, --prohibit=<i,j,k>
-f, --force=<i,j,k>
-E, --energyOnly
-I, --noisolate
-F, --mfold=<P,W,MAX> (defaults to 5,3,100)
-c, --constraints=<name of constraints file>
-b, --basepairs=<name of basepairs file>
    --temperature=<hybridization temperature>

Obscure options:
    --allpairs
    --maxloop=<maximum bulge/interior loop size> (defaults to 30)
    --nodangle
    --prefilter=<filter value>

EOF
print 'Report bugs to markhn@rpi.edu', "\n";
    exit;
}

my ($tMin, $tInc, $tMax, $temp) = (0, 1, 100, 37);
my ($energyOnly, @prohibit, @force);
my %options;

GetOptions \%options, 'version|V' => sub { version('hybrid-2s.pl') }, 'help|h' => sub { usage() }, 'NA|n=s', 'tmin|t=f' => \$tMin, 'tinc|i=f' => \$tInc, 'tmax|T=f' => \$tMax, 'allpairs', 'sodium|N=f', 'magnesium|M=f', 'polymer|p', 'prohibit|r=s' => \@prohibit, 'force|f=s' => \@force, 'energyOnly|E' => \$energyOnly, 'noisolate|I', 'mfold|F:s', 'constraints|c:s', 'basepairs|b=s', 'maxloop=i', 'nodangle', 'prefilter=s', 'temperature=f' => \$temp or die $!;

my @args = ('--tmin' => $temp, '--tmax' => $temp);
foreach (keys %options) {
    if ($_ eq 'allpairs' or $_ eq 'nodangle' or $_ eq 'polymer' or $_ eq 'noisolate') {
	push @args, "--$_";
    } else {
	push @args, "--$_";
	push @args, $options{$_} if defined $options{$_};
    }
}

my ($file1, $file2) = @ARGV;
unless (defined $file1 and defined $file2) {
    print STDERR "Error: data not specified\nRun 'hybrid-2s.pl -h' for help\n";
    exit 1;
}
my ($prefix1, $prefix2);
($prefix1 = $file1) =~ s/\.seq$//;
($prefix2 = $file2) =~ s/\.seq$//;
my $prefix = "$prefix1-$prefix2";

system('hybrid-min', @args, $file1, $file2) == 0 or die $!;

my (@struct, @dG, @dH, @dS);

my $len1;
open IN, "<$prefix.ct" or die $!;
while (<IN>) {
    chomp;
    my ($len, $dG) = /^(.+)\tdG = (.+?)\t/;
    push @dG, $dG;
    my @ss;
    for (my $i = 0; $i < $len; ++$i) {
	my $line = <IN>;
	$ss[$i] = (split "\t", $line)[4] ? 0 : 1;
	$len1 = $i if (split "\t", $line)[2] == 0;
    }
    push @struct, \@ss;
}
close IN or die $!;

my $suffix = (defined $options{NA} and $options{NA} eq 'DNA') ? 'DHD' : 'DH';
open IN, "ct-energy -s$suffix $prefix.ct|" or die $!;
while (<IN>) {
    chomp;
    s/inf/999999/;
    push @dH, $_;
}
close IN or die $!;

for (my $i = 0; $i < @dG; ++$i) {
    if ($dG[$i] > 500) {
	push @dS, -999999;
    } else {
	push @dS, ($dH[$i] - $dG[$i]) / (273.15 + $temp);
    }
}

open IN, "<$prefix.$temp.ext" or die $!;
while (<IN>) {
    /^1\t(\d+)/ and $len1 = $1;
    /^2\t(\d+)/ and $struct[0][$1 + $len1 - 1] = 0;
}
close IN or die $!;

open OUT, ">$prefix.dG" or die $!;
print OUT "#T\t-RT ln Z\tZ\n";
for (my $t = $tMin; $t <= $tMax; $t += $tInc) {
    my $z = 0;
    my @p1 = (0) x @{$struct[0]};
    my @p2 = (0) x (@{$struct[0]} - 1);
    for (my $i = 0; $i < @dG; ++$i) {
	$z += exp(-($dH[$i] - (273.15 + $t) * $dS[$i]) / R / (273.15 + $t));
    }
    for (my $i = 0; $i < @dG; ++$i) {
	for (my $j = 0; $j < @{$struct[$i]}; ++$j) {
	    $p1[$j] += exp(-($dH[$i] - (273.15 + $t) * $dS[$i]) / R / (273.15 + $t)) / $z if $struct[$i][$j];
	    $p2[$j] += exp(-($dH[$i] - (273.15 + $t) * $dS[$i]) / R / (273.15 + $t)) / $z if $j < @{$struct[$i]} - 1 and $struct[$i][$j] and $struct[$i][$j + 1];
	}
    }
    if ($z == 0) {
	printf OUT "%g\t%s\t%g\n", $t, 'inf', 0;
    } else {
	printf OUT "%g\t%g\t%g\n", $t, - R * (273.15 + $t) * log($z), $z;
    }

    unless ($energyOnly) {
	open OUT2, ">$prefix.$t.ext" or die $!;
	print OUT2 "sequence\ti/j\tP(i is SS)\tP(i is SS and i+1 is SS)\n";
	for (my $i = 0; $i < $len1 - 1; ++$i) {
	    printf OUT2 "1\t%d\t%g\t%g\n", $i + 1, $p1[$i], $p2[$i];
	}
	printf OUT2 "1\t%d\t%g\n", $len1, $p1[$len1 - 1];
	for (my $i = $len1; $i < @{$struct[0]} - 1; ++$i) {
	    printf OUT2 "2\t%d\t%g\t%g\n", $i + 1 - $len1, $p1[$i], $p2[$i];
	}
	printf OUT2 "2\t%d\t%g\n", @{$struct[0]} - $len1, $p1[@{$struct[0]} - 1];
	close OUT2 or die $!;
    }
}
close OUT or die $!;
